vncviewer/touch.cxx

```diff
@@ -0,0 +1,627 @@
+        fakeButtonEvent(true, Button4, ev);
+        fakeButtonEvent(false, Button4, ev);
+        lastDistance += ZOOMSENS;
+      }
+      while ((ev.detail - lastDistance) < -ZOOMSENS) {
+        fakeButtonEvent(true, Button5, ev);
+        fakeButtonEvent(false, Button5, ev);
+        lastDistance -= ZOOMSENS;
+      }
+    }
+    break;
+
+  case GH_GestureEnd:
+    switch (ev.gesture) {
+    case GH_LEFTBTN:
+      vlog.info("Got GH_GestureEnd(GH_LEFTBTN)");
+      fakeButtonEvent(false, Button1, ev);
+      break;
+    case GH_MIDDLEBTN:
+      vlog.info("Got GH_GestureEnd(GH_MIDDLEBTN)");
+      fakeButtonEvent(false, Button2, ev);
+      break;
+    case GH_RIGHTBTN:
+      vlog.info("Got GH_GestureEnd(GH_RIGHTBTN)");
+      fakeButtonEvent(false, Button3, ev);
+      break;
+    case GH_VSCROLL:
+      vlog.info("Got GH_GestureEnd(GH_VSCROLL)");
+      break;
+    case GH_HSCROLL:
+      vlog.info("Got GH_GestureEnd(GH_HSCROLL)");
+      break;
+    case GH_ZOOM:
+      vlog.info("Got GH_GestureEnd(GH_ZOOM)");
+      fakeKeyEvent(false, XK_Control_L, ev);
+      break;
+    }
+    break;
+  }
+}
+
+static int handleXinputEvent(void *event, void *data)
+{
+  XEvent *xevent = (XEvent*)event;
+
+  if (xevent->type == MapNotify) {
+    XIEventMask eventmask;
+    unsigned char flags[XIMaskLen(XI_LASTEVENT)] = { 0 };
+
+    eventmask.deviceid = XIAllMasterDevices;
+    eventmask.mask_len = sizeof(flags);
+    eventmask.mask = flags;
+
+    XISetMask(flags, XI_ButtonPress);
+    XISetMask(flags, XI_Motion);
+    XISetMask(flags, XI_ButtonRelease);
+
+    XISetMask(flags, XI_TouchBegin);
+    XISetMask(flags, XI_TouchUpdate);
+    XISetMask(flags, XI_TouchEnd);
+
+    XISelectEvents(fl_display, xevent->xmap.window, &eventmask, 1);
+
+    handlers[xevent->xmap.window] = new FLTKGestureHandler(xevent->xmap.window);
+
+    // Fall through as we don't want to interfere with whatever someone
+    // else might want to do with this event
+
+  } else if (xevent->type == UnmapNotify) {
+    handlers.erase(xevent->xunmap.window);
+  } else if (xevent->type == DestroyNotify) {
+    handlers.erase(xevent->xdestroywindow.window);
+  } else if (xevent->type == GenericEvent) {
+    if (xevent->xgeneric.extension == xi_major) {
+      XIDeviceEvent *devev;
+
+      if (!XGetEventData(fl_display, &xevent->xcookie)) {
+        vlog.error("Failed to get event data for X Input event");
+        return 1;
+      }
+
+      devev = (XIDeviceEvent*)xevent->xcookie.data;
+
+      if (handlers.count(devev->event) == 0) {
+        vlog.error("X Input event for unknown window");
+        XFreeEventData(fl_display, &xevent->xcookie);
+        return 1;
+      }
+
+      // FLTK doesn't understand X Input events, and we've stopped
+      // delivery of Core events by enabling the X Input ones. Make
+      // FLTK happy by faking Core events based on the X Input ones.
+
+      if (xevent->xgeneric.evtype == XI_TouchBegin) {
```

Comment:
A comment about why this is here would be nice.

Reply:
91067c84